state/application.go

```diff
@@ -3124,6 +3132,94 @@ func allUnits(st *State, application string) (units []*Unit, err error) {
 	return units, nil
 }
 
+// GetUnitAttachmentInfos returns storage attachment info for units not yet provisioned,
+// based on DesiredScale and UnitCount.
+//
+// This is called by the same worker loop that deploy application && updates scales.
+// So consistency checks can be avoided since provisioning and scale updates are interleaved.
+func (a *Application) GetUnitAttachmentInfos() (unitAttachmentInfos []UnitAttachmentInfo, err error) {
+	// Skips if scaling down or already at desired scale (except initial 1-unit deploy).
+	// Note: Attaching storage is not allowed during deployment if DesiredScale > 1, so that case is ignored.
+	if a.doc.UnitCount >= a.doc.DesiredScale && !(a.doc.UnitCount == 1 && a.doc.DesiredScale == 1) {
+		return unitAttachmentInfos, nil
+	}
+
+	unitIds := []string{}
+
+	start := a.doc.UnitCount
+	if a.doc.UnitCount == 1 && a.doc.DesiredScale == 1 {
+		start = 0
+	}
+	for i := start; i < a.doc.DesiredScale; i++ {
+		unitIds = append(unitIds, strconv.Itoa(i))
+	}
+
+	idReg := strings.Join(unitIds, "|")
+	storageAttachmentDocs, err := getstorageAttachmentDocs(
+		a.st.db(),
+		bson.M{
+			"unitid": bson.RegEx{
+				Pattern: fmt.Sprintf(
+					`^%s/(?:%s)$`,
+					regexp.QuoteMeta(a.doc.Name),
+					idReg,
+				),
+				Options: "",
+			},
+		},
+	)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	var storageIds []string
+	storageAttachmentDocByStorageId := make(map[string]storageAttachmentDoc)
+	for _, stDoc := range storageAttachmentDocs {
+		storageIds = append(
+			storageIds,
+			stDoc.StorageInstance,
+		)
+		storageAttachmentDocByStorageId[stDoc.StorageInstance] = stDoc
+	}
+
+	volumeDocs, err := getVolumeDocs(
+		a.st.db(),
+		bson.M{
+			"info":      bson.M{"$ne": nil},
+			"storageid": bson.M{"$in": storageIds},
+		},
+	)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	for _, volDoc := range volumeDocs {
+		if stDoc, ok := storageAttachmentDocByStorageId[volDoc.StorageId]; ok {
+			unitAttachmentInfos = append(
+				unitAttachmentInfos,
+				UnitAttachmentInfo{
+					Unit:      stDoc.Unit,
+					StorageId: volDoc.StorageId,
+					VolumeId:  volDoc.Info.VolumeId,
+				},
+			)
+		}
+	}
+	return unitAttachmentInfos, nil
+}
+
+func getstorageAttachmentDocs(db Database, query interface{}) ([]storageAttachmentDoc, error) {
+	coll, cleanup := db.GetCollection(storageAttachmentsC)
+	defer cleanup()
+
+	var docs []storageAttachmentDoc
+	err := coll.Find(query).All(&docs)
```

Comment:
We are only interested in the storage id so do something similar to this which is used in `cleanupAttachmentsForDyingStorage`
The 'fields` and `Select()` avoid loading the entire doc. Use "storageid" instead of "unitid" of course.
```
	fields := bson.D{{"unitid", 1}}
	iter := coll.Find(blah).Select(fields).Iter()
```

Reply:
Ack to adding `Select`, but we need both unit id & storage id here.